File: packages/sdk/src/socket.ts
import { io, type Socket } from 'socket.io-client'
import type { ClientEventMap, ServerEventMap } from './types'

export interface ShadowSocketOptions {
  /** Shadow server base URL (e.g. "https://shadowob.shadowob.com") */
  serverUrl: string
  /** JWT token for authentication */
  token: string
  /** Socket.IO transports (default: ['websocket']) */
  transports?: string[]
  /** Auto-reconnect on disconnect (default: true) */
  autoReconnect?: boolean
  /** Reconnection delay in ms (default: 1000) */
  reconnectionDelay?: number
}

type ServerEventName = keyof ServerEventMap
type ServerEventHandler<E extends ServerEventName> = ServerEventMap[E]

/**
 * Shadow real-time event listener.
 *
 * Wraps Socket.IO with strongly-typed events that match the Shadow server
 * gateway broadcasts. Provides channel/thread room management and
 * convenience methods for sending messages and typing indicators.
 */
export class ShadowSocket {
  private socket: Socket
  private _connected = false

  constructor(options: ShadowSocketOptions) {
    this.socket = io(options.serverUrl, {
      auth: { token: options.token },
      transports: options.transports ?? ['websocket'],
      autoConnect: false,
      reconnection: options.autoReconnect ?? true,
      reconnectionDelay: options.reconnectionDelay ?? 1000,
    })

    this.socket.on('connect', () => {
      this._connected = true
    })
    this.socket.on('disconnect', () => {
      this._connected = false
    })
  }

  /** Whether the socket is currently connected */
  get connected(): boolean {
    return this._connected
  }

  /** The underlying Socket.IO socket instance */
  get raw(): Socket {
    return this.socket
  }

  // ── Connection lifecycle ──────────────────────────────────────────────

  /** Connect to the Shadow server */
  connect(): void {
    if (!this.socket.connected) {
      this.socket.connect()
    }
  }

  /** Disconnect from the Shadow server */
  disconnect(): void {
    this.socket.disconnect()
  }

  /** Wait until the socket is connected (resolves immediately if already connected) */
  waitForConnect(timeoutMs = 5000): Promise<void> {
    if (this.socket.connected) return Promise.resolve()
    return new Promise<void>((resolve, reject) => {
      const timer = setTimeout(() => {
        reject(new Error(`Socket connect timeout after ${timeoutMs}ms`))
      }, timeoutMs)
      this.socket.once('connect', () => {
        clearTimeout(timer)
        resolve()
      })
    })
  }

  // ── Typed event listeners ─────────────────────────────────────────────

  /** Listen for a server event */
  on<E extends ServerEventName>(event: E, handler: ServerEventHandler<E>): this {
    this.socket.on(event as string, handler as (...args: unknown[]) => void)
    return this
  }

  /** Listen for a server event (one-time) */
  once<E extends ServerEventName>(event: E, handler: ServerEventHandler<E>): this {
    this.socket.once(event as string, handler as (...args: unknown[]) => void)
    return this
  }

  /** Remove a specific event listener */
  off<E extends ServerEventName>(event: E, handler: ServerEventHandler<E>): this {
    this.socket.off(event as string, handler as (...args: unknown[]) => void)
    return this
  }

  /** Remove all listeners for an event or all events */
  removeAllListeners(event?: ServerEventName): this {
    if (event) {
      this.socket.removeAllListeners(event)
    } else {
      this.socket.removeAllListeners()
    }
    return this
  }

  // ── Connection event listeners ────────────────────────────────────────

  /** Listen for raw connection events (connect, disconnect, connect_error) */
  onConnect(handler: () => void): this {
    this.socket.on('connect', handler)
    return this
  }

  onDisconnect(handler: (reason: string) => void): this {
    this.socket.on('disconnect', handler)
    return this
  }

  onConnectError(handler: (error: Error) => void): this {
    this.socket.on('connect_error', handler)
    return this
  }

  // ── Room management ───────────────────────────────────────────────────

  /** Join a channel room to receive its messages and events */
  joinChannel(channelId: string): Promise<{ ok: boolean }> {
    return new Promise((resolve) => {
      this.socket.emit(
        'channel:join' satisfies keyof ClientEventMap,
        { channelId },
        (res: { ok: boolean }) => {
          resolve(res ?? { ok: true })
        },
      )
    })
  }

  /** Leave a channel room */
  leaveChannel(channelId: string): void {
    this.socket.emit('channel:leave' satisfies keyof ClientEventMap, { channelId })
  }

  // ── Client actions ────────────────────────────────────────────────────

  /** Send a message via WebSocket (text-only; for file attachments use REST) */
  sendMessage(data: {
    channelId: string
    content: string
    threadId?: string
    replyToId?: string
  }): void {
    this.socket.emit('message:send' satisfies keyof ClientEventMap, data)
  }

  /** Send a typing indicator */
  sendTyping(channelId: string): void {
    this.socket.emit('message:typing' satisfies keyof ClientEventMap, { channelId })
  }

  /** Update user presence status */
  updatePresence(status: 'online' | 'idle' | 'dnd' | 'offline'): void {
    this.socket.emit('presence:update' satisfies keyof ClientEventMap, { status })
  }

  /** Update activity status in a channel (e.g. 'thinking', 'working', null) */
  updateActivity(channelId: string, activity: string | null): void {
    this.socket.emit('presence:activity' satisfies keyof ClientEventMap, { channelId, activity })
  }

  // ── DM actions ────────────────────────────────────────────────────────

  /** Join a DM channel room */
  joinDmChannel(dmChannelId: string): void {
    this.socket.emit('dm:join' as string, { dmChannelId })
  }

  /** Leave a DM channel room */
  leaveDmChannel(dmChannelId: string): void {
    this.socket.emit('dm:leave' as string, { dmChannelId })
  }

  /** Send a DM message via WebSocket */
  sendDmMessage(data: {
    dmChannelId: string
    content: string
    replyToId?: string
    metadata?: Record<string, unknown>
  }): void {
    this.socket.emit('dm:send' as string, data)
  }

  /** Send a DM typing indicator */
  sendDmTyping(dmChannelId: string): void {
    this.socket.emit('dm:typing' as string, { dmChannelId })
  }
}

File: packages/sdk-python/shadowob_sdk/socket.py
"""Shadow real-time Socket.IO client — Python edition.

Mirrors the TypeScript ``ShadowSocket`` class so that every JS SDK socket
method has a Python equivalent with identical semantics.
"""

from __future__ import annotations

from typing import Any, Callable

import socketio


class ShadowSocket:
    """Real-time event listener for the Shadow server via Socket.IO."""

    def __init__(
        self,
        server_url: str,
        token: str,
        *,
        transports: list[str] | None = None,
        auto_reconnect: bool = True,
        reconnection_delay: float = 1.0,
    ) -> None:
        self._server_url = server_url.rstrip("/")
        self._token = token
        self._connected = False
        self._sio = socketio.Client(
            reconnection=auto_reconnect,
            reconnection_delay=reconnection_delay,
        )

        # Internal connection tracking
        @self._sio.event
        def connect() -> None:
            self._connected = True

        @self._sio.event
        def disconnect() -> None:
            self._connected = False

    @property
    def connected(self) -> bool:
        return self._connected

    @property
    def raw(self) -> socketio.Client:
        return self._sio

    # ── Connection lifecycle ─────────────────────────────────────────────

    def connect(self) -> None:
        if not self._sio.connected:
            self._sio.connect(
                self._server_url,
                auth={"token": self._token},
                transports=["websocket"],
            )

    def disconnect(self) -> None:
        if self._sio.connected:
            self._sio.disconnect()

    def wait(self) -> None:
        """Block until the connection is closed (useful for long-running bots)."""
        self._sio.wait()

    # ── Event listeners ──────────────────────────────────────────────────

    def on(self, event: str, handler: Callable[..., Any]) -> "ShadowSocket":
        self._sio.on(event, handler)
        return self

    def off(self, event: str, handler: Callable[..., Any] | None = None) -> "ShadowSocket":
        # python-socketio uses on() to register; to remove we access handlers directly
        if handler and event in self._sio.handlers:
            fns = self._sio.handlers[event]
            self._sio.handlers[event] = [f for f in fns if f is not handler]
        elif event in self._sio.handlers:
            del self._sio.handlers[event]
        return self

    def on_connect(self, handler: Callable[[], None]) -> "ShadowSocket":
        self._sio.on("connect", handler)
        return self

    def on_disconnect(self, handler: Callable[[str], None]) -> "ShadowSocket":
        self._sio.on("disconnect", handler)
        return self

    # ── Room management ──────────────────────────────────────────────────

    def join_channel(self, channel_id: str) -> dict[str, Any]:
        return self._sio.call("channel:join", {"channelId": channel_id})

    def leave_channel(self, channel_id: str) -> None:
        self._sio.emit("channel:leave", {"channelId": channel_id})

    def join_voice_channel(
        self,
        channel_id: str,
        *,
        client_id: str | None = None,
        muted: bool | None = None,
        deafened: bool | None = None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {"channelId": channel_id}
        if client_id is not None:
            payload["clientId"] = client_id
        if muted is not None:
            payload["muted"] = muted
        if deafened is not None:
            payload["deafened"] = deafened
        return self._sio.call("voice:join", payload)

    def leave_voice_channel(self, channel_id: str) -> dict[str, Any]:
        return self._sio.call("voice:leave", {"channelId": channel_id})

    def update_voice_state(
        self,
        channel_id: str,
        *,
        muted: bool | None = None,
        deafened: bool | None = None,
        speaking: bool | None = None,
        screen_sharing: bool | None = None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {"channelId": channel_id}
        if muted is not None:
            payload["muted"] = muted
        if deafened is not None:
            payload["deafened"] = deafened
        if speaking is not None:
            payload["speaking"] = speaking
        if screen_sharing is not None:
            payload["screenSharing"] = screen_sharing
        return self._sio.call("voice:state:update", payload)

    def send_voice_heartbeat(self, channel_id: str) -> None:
        self._sio.emit("voice:heartbeat", {"channelId": channel_id})

    # ── Client actions ───────────────────────────────────────────────────

    def send_message(
        self,
        channel_id: str,
        content: str,
        *,
        thread_id: str | None = None,
        reply_to_id: str | None = None,
        mentions: list[dict[str, Any]] | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> None:
        data: dict[str, Any] = {"channelId": channel_id, "content": content}
        if thread_id:
            data["threadId"] = thread_id
        if reply_to_id:
            data["replyToId"] = reply_to_id
        if mentions is not None:
            data["mentions"] = mentions
        if metadata is not None:
            data["metadata"] = metadata
        self._sio.emit("message:send", data)

    def send_typing(self, channel_id: str) -> None:
        self._sio.emit("message:typing", {"channelId": channel_id})

    def update_presence(
        self, status: str  # 'online' | 'idle' | 'dnd' | 'offline'
    ) -> None:
        self._sio.emit("presence:update", {"status": status})

    def update_activity(
        self, channel_id: str, activity: str | None
    ) -> None:
        self._sio.emit(
            "presence:activity", {"channelId": channel_id, "activity": activity}
        )
